File: go/weightedgraph/lazy_prim.go
package weightedgraph

import (
	// "fmt"
)

type LazyPrimMST struct {
	graph Graph
	heap *MinHeap
	marked []bool
	mst []Edge
	mstWeight float32
}


func NewLazyPrim(graph Graph) *LazyPrimMST {
	mst := &LazyPrimMST{graph: graph,}
	heap, _ := NewMinHeap(graph.E())
	mst.heap = heap
	mst.marked = make([]bool, graph.V())
	mst.mst = []Edge{}

	// Lazy Prim
	mst.lazyPrim()
	return mst
}


func (this *LazyPrimMST) lazyPrim() {
	// 任选一个节点开始我们的切分过程
	this.visit(0)

	for ; !this.heap.IsEmpty(); {
		edge, _ := this.heap.RemoveMin()

		// 如果当前这条权值最小的边已经不再是一条横切边了
		if this.marked[edge.V()] == this.marked[edge.W()] {
			continue
		}
		this.mst = append(this.mst, edge)
		if this.marked[edge.V()] {
			this.visit(edge.W())
		} else {
			this.visit(edge.V())
		}
	}

	for i:=0; i < len(this.mst); i++ {
		this.mstWeight += this.mst[i].Weight()
	}
}

func (this *LazyPrimMST) visit(v int) {
	if this.marked[v] {
		return
	}

	this.marked[v] = true
	vertexs, _ := this.graph.Vertexs(v)

	for i := 0; i < len(vertexs); i++ {
		vertex := vertexs[i]
		// this.marked[other] == false 说明这条边是一条横切边，入堆
		if other, _ := vertex.Other(v);!this.marked[other] {
			this.heap.Add(vertex)
		}
	}
	// fmt.Println(this.heap.GetSize())
}

func (this *LazyPrimMST) MSTEdges() []Edge {
	return this.mst
}

func (this *LazyPrimMST) Result() float32 {
	return this.mstWeight
}

File: js/graphMST/lazyPrim.js
const PQ = require('./minHeap');

class LazyPrim {
  constructor(graph) {    
    this.graph = graph;
    this.marked = Array(graph.V());
    this.mst = [];
    this.mstWeight = 0;
    this.heap = new PQ(this.graph.E())

    for (let i = 0; i < this.marked.length; i++) {
      this.marked[i] = false;
    }

    this.lazyPrim();
  }
  
  lazyPrim() {
    this.visit(0);
    while(!this.heap.IsEmpty()) {
      const edge = this.heap.ExtractMin();
      if (this.marked[edge.V()] === this.marked[edge.W()]) {
        continue;
      }
      this.mst.push(edge);
      this.mstWeight += edge.Weight();
      if (this.marked[edge.V()]) {
        this.visit(edge.W());
      } else {
        this.visit(edge.V());
      }
    }
  }

  visit(v) {
    if (this.marked[v]) {
      return;
    }

    this.marked[v] = true;
    const adjs = this.graph.Adj(v);
    for (let i = 0; i < adjs.length; i++) {
      const w = adjs[i].Other(v);
      if (!this.marked[w]) {
        this.heap.Insert(adjs[i]);
      }
    }
  }

  MSTEdges() {
    return this.mst;
  }

  Result() {
    return this.mstWeight;
  }
}

module.exports = LazyPrim;
